File: Sources/QuillKit/Hotkey/UndoChord.swift
import CoreGraphics
import Foundation

/// ⌥⌫ — "take back what you just inserted" — as a decision rather than a
/// keystroke.
///
/// Pure, and separated from the tap for the same reason the gesture grammar is:
/// this is the part that decides whether Quill deletes text out of somebody
/// else's document, and it has to be checkable without a keyboard.
///
/// It is not part of `HotkeyStateMachine` because it is not a gesture — there is
/// no hold, no double-tap and no timer in it. It does have to know what the
/// gesture machine is doing, which is why the state comes in as an argument.
public enum UndoChord {

    /// The key labelled ⌫. Shared with `SyntheticKeyboard` so the code that
    /// swallows the chord and the code that puts it back cannot drift apart.
    public static let keyCode: UInt16 = SyntheticKeyboard.keyBackspace
    public static let flags: CGEventFlags = .maskAlternate

    /// Whether this keystroke should be claimed by Quill instead of reaching the
    /// app.
    ///
    /// Every clause is a way of not firing, which is the point: ⌥⌫ deletes the
    /// previous word in every macOS text field, and an override that fires when
    /// it should not is worse than no override at all.
    /// How long after the trigger went down a ⌫ can still be the other half of a
    /// chord rather than a key pressed during a dictation.
    ///
    /// The arm delay is 120 ms, and a deliberate two-key reach — thumb on right
    /// Option, finger up to Delete — takes longer than that more often than not.
    /// Without this the feature fires or does not fire depending on how fast the
    /// user chords, which reads as "it works sometimes". Above it, nothing: 400 ms
    /// of held trigger with a ⌫ at the end is not a sentence anyone was speaking.
    public static let chordWindow: TimeInterval = 0.4

    public static func claims(keyCode: UInt16,
                              flags: CGEventFlags,
                              gesture: HotkeyStateMachine.State,
                              triggerHeldFor: TimeInterval?,
                              hasInsertion: Bool) -> Bool {
        guard hasInsertion else { return false }
        guard keyCode == Self.keyCode else { return false }
        // Exactly Option. ⇧⌥⌫ and ⌘⌥⌫ are other people's bindings and are left
        // alone; `chordMask` is what strips the fn and numeric-pad bits that
        // arrive set on a built-in keyboard.
        guard flags.intersection(HotkeyBinding.chordMask) == Self.flags else { return false }

        switch gesture {
        case .idle:
            return true
        case .armed:
            // The dictation key may itself be an Option — right Option is the
            // default. Pressing ⌥⌫ therefore arms a speculation with the modifier
            // and delivers the ⌫ a few milliseconds later, and without this clause
            // the chord would never work for anyone using the shipped binding.
            // Nobody starts a dictation and types a backspace inside the 120ms arm
            // delay, so this is the chord and not a gesture.
            return true
        case .holding:
            // Same chord, slightly slower hand. The arm timer moved the machine on
            // at 120 ms whether or not the user was doing anything, so "we are in
            // .holding" does not yet mean "the user is speaking" — for the first
            // few hundred milliseconds it means "the trigger is down", which is
            // exactly what the first half of ⌥⌫ looks like.
            //
            // Past the window the original reading holds: the user is speaking, a
            // ⌫ means "cancel this dictation", and claiming it would delete the
            // previous sentence as well as binning the current one.
            guard let triggerHeldFor, triggerHeldFor <= chordWindow else { return false }
            return true
        case .handsFree:
            // Nothing is held here, so there is no chord to be halfway through —
            // and unlike a hold, hands-free is expected to survive typing. A ⌫ is
            // the user editing, in a field a live dictation is writing into.
            return false
        }
    }
}

File: src/core/hotkey/undoChord.ts
import type { HotkeyState } from './stateMachine';

/// "Take back what you just inserted" — as a decision rather than a keystroke.
///
/// Pure, and separated from the hook for the same reason the gesture grammar
/// is: this is the part that decides whether Quill deletes text out of somebody
/// else's document, and it has to be checkable without a keyboard.
///
/// # What changed from the macOS build, and why
///
/// There, the chord was ⌥⌫ and the event tap SWALLOWED it, because ⌥⌫ already
/// means "delete the previous word" in every macOS text field. Swallowing is
/// the part that cannot be done here: `uiohook` observes the keyboard, it does
/// not sit in the delivery path, so a key it sees has already reached the
/// focused app. Overriding a standard binding without being able to swallow it
/// would mean the app deletes a word AND Quill deletes its sentence — two
/// deletions for one keystroke, which is worse than not having the feature.
///
/// So on Windows and Linux the chord is an Electron **global accelerator**,
/// Ctrl+Alt+Z by default, which the OS does deliver exclusively to us. That
/// removes the swallow problem by removing the collision: Ctrl+Alt+Z is not
/// bound by default anywhere, so nothing is being overridden. On macOS the
/// default stays Alt+Backspace, registered the same way.
///
/// What survives unchanged is the interesting half — the rule about WHEN the
/// chord may fire, which is what stops it deleting the previous sentence
/// because a key happened to be pressed mid-dictation.

/// How long after the trigger went down a chord press can still be the other
/// half of a reach rather than a key pressed during a dictation.
///
/// A backstop, not the test.
///
/// The question in `holding` is "is the user speaking, or reaching for the
/// other half of a chord", and the honest answer to it is `heardSpeech` —
/// whether the microphone has picked anything up. A clock was tried first and
/// the measurement killed it: driving the real chord through the real hook,
/// with a 500 ms gap between the two keys, the second key arrived 1.075 s after
/// the trigger went down. A human reaching across a keyboard does not beat a
/// shell script. Any window a careful press falls outside of turns the feature
/// back into "sometimes".
///
/// So this only catches the case where the audio never reports at all — a dead
/// microphone, a device that went away — and it is set well beyond any reach
/// and well inside a real dictation.
export const CHORD_WINDOW_SECONDS = 5;

export interface UndoChordQuery {
  gesture: HotkeyState;
  /// Seconds since the trigger went down, or null when it is not down.
  triggerHeldFor: number | null;
  /// Whether the microphone has heard anything above the noise floor during the
  /// gesture currently in flight.
  heardSpeech: boolean;
  /// Whether there is an insertion to take back at all.
  hasInsertion: boolean;
}

/// Whether the chord should take back Quill's last insertion.
///
/// Every clause is a way of not firing, which is the point.
export function undoChordClaims(query: UndoChordQuery): boolean {
  if (!query.hasInsertion) return false;

  switch (query.gesture.kind) {
    case 'idle':
      return true;
    case 'armed':
      // The dictation key may itself be part of the reach. Nobody starts a
      // dictation and presses the undo chord inside the 120 ms arm delay, so
      // this is the chord and not a gesture.
      return true;
    case 'holding':
      // Same chord, slightly slower hand. The arm timer moved the machine on at
      // 120 ms whether or not the user was doing anything, so "we are in
      // holding" does not yet mean "the user is speaking" — for the first few
      // hundred milliseconds it means "the trigger is down".
      //
      // Once anything has been heard the original reading holds: the user is
      // speaking, and claiming the chord would delete the previous sentence as
      // well as binning the current one.
      if (query.heardSpeech) return false;
      if (query.triggerHeldFor === null) return false;
      return query.triggerHeldFor <= CHORD_WINDOW_SECONDS;
    case 'handsFree':
      // Nothing is held here, so there is no chord to be halfway through — and
      // unlike a hold, hands-free is expected to survive typing. A keystroke is
      // the user editing, in a field a live dictation is writing into.
      return false;
  }
}
